### PR TITLE
Migrate Get/Grant/Revoke/Test-CPermission from Carbon

### PR DESCRIPTION
**Pull request by splatteredbits**

### CODE REVIEW

**Comment by splatteredbits** [comment]
This is ready for a review.

File: Carbon.Permissions/Carbon.Permissions.psd1 [hunk 1]
```diff
@@ -102,21 +107,25 @@
...
+            ProjectUri = 'https://github.com/webmd-health-services/Carbon.Permissions'
 
             # A URL to an icon representing this module.
             # IconUri = ''
 
             Prerelease = ''
 
             # ReleaseNotes of this module
-            ReleaseNotes = 'https://github.com/GITHUB_ORGANIZATION_NAME/MODULE_NAME/blob/main/CHANGELOG.md'
+            ReleaseNotes = 'https://github.com/webm-health-services/Carbon.Permissions/blob/main/CHANGELOG.md'
```
Comment (decoyjoe) [inline Carbon.Permissions/Carbon.Permissions.psd1]: Typo: Missing the `d` on `webmd`

File: Carbon.Permissions/Carbon.Permissions.psm1 [hunk 2]
```diff
@@ -16,12 +20,24 @@
...
 # module in development has its functions in the Functions directory.
 $moduleRoot = $PSScriptRoot
 
-# Store each of your module's functions in its own file in the Functions 
-# directory. On the build server, your module's functions will be appended to 
+$psModulesRoot = Join-Path -Path $PSScriptRoot -ChildPath 'Modules' -Resolve
+Import-Module -Name (Join-Path -Path $psModulesRoot -ChildPath 'Carbon.Core') `
+              -Function @('Get-CPathProvider')
+Import-Module -Name (Join-Path -Path $psModulesRoot -ChildPath 'Carbon.Accounts') `
+              -Function @('Resolve-CIdentityName', 'Test-CIdentity')
```
Comment (decoyjoe) [inline Carbon.Permissions/Carbon.Permissions.psm1]: Add `-Verbose:$false`

File: Carbon.Permissions/Carbon.Permissions.psm1 [hunk 3]
```diff
@@ -16,12 +20,24 @@
...
+
+if (-not (Test-Path -Path 'variable:IsWindows'))
+{
+    $IsWindows = $true
+    $IsMacOS = $IsLinux = $true
```
Comment (decoyjoe) [inline Carbon.Permissions/Carbon.Permissions.psm1]: This should be `$false`

File: Carbon.Permissions/Functions/Get-CPermission.ps1 [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    Returns `System.Security.AccessControl.CryptoKeyAccesRule` objects for certificate's
+    `Cert:\LocalMachine\1234567890ABCDEF1234567890ABCDEF12345678` private key/key container. If it doesn't have a
+    private key, `$null` is returned.
+    #>
+    [CmdletBinding()]
+    [OutputType([System.Security.AccessControl.AccessRule])]
+    param(
+        # The path whose permissions (i.e. access control rules) to return. File system, registry, or certificate paths
+        # supported. Wildcards supported. For certificate private keys, pass a certificate provider path, e.g. `cert:`.
```
Comment (decoyjoe) [inline Carbon.Permissions/Functions/Get-CPermission.ps1:70]: Does this module work cross-platform? I think the `Cert:\` drive is only available on Windows.

File: Carbon.Permissions/Functions/Test-CPermission.ps1 [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+        # The propagation flags to check for. Default is to ignore propagation flags when testing the permission.
+        [PropagationFlags] $PropagationFlag,
+
+        # Include inherited permissions in the check.
+        [switch] $Inherited,
+
+        # Check for the exact permissions, inheritance flags, and propagation flags, i.e. make sure the identity has
+        # *only* the permissions you specify.
+        [switch] $Exact
```
Comment (decoyjoe) [inline Carbon.Permissions/Functions/Test-CPermission.ps1]: In other modules I think we use `-Strict` for this?

File: Carbon.Permissions/en-US/about_Carbon.Permissions.help.txt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+TOPIC
+    about_Carbon.Permissions
+
+SHORT DESCRIPTION
+    Carbon.Permissions is a PowerShell module for...
+
+LONG DESCRIPTION
+    The Carbon.Permissions PowerShell module...
```
Comment (decoyjoe) [inline Carbon.Permissions/en-US/about_Carbon.Permissions.help.txt]: Missing descriptions.

File: TODO.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
+
+# Template-PSModules-OSS-Apache2.0-AppVeyor
```
Comment (decoyjoe) [inline TODO.md]: Ok to remove this file?

File: README.md [hunk 8]
```diff
@@ -1,36 +1,26 @@
...
-  * Windows PowerShell 5.1/.NET 4.6.2
-  * Windows PowerShell 5.1/.NET 4.8
-  * PowerShell 6.2 on Windows
-  * PowerShell 7.1 on Windows
-  * PowerShell 7.2 on Windows
-  * PowerShell 7.1 on macOS
-  * PowerShell 7.2 on Ubuntu
+# Overview
+
+The "Carbon.Permissions" module...
```
Comment (decoyjoe) [inline README.md]: Missing description.

File: README.md [hunk 9]
```diff
@@ -1,36 +1,26 @@
...
+
+# System Requirements
+
+* Windows PowerShell 5.1 and .NET 4.6.1+
+* PowerShell Core 6+
```
Comment (decoyjoe) [inline README.md]: I think this module is Windows only? I.e. `PowerShell Core 6+ on Windows`

**Review by decoyjoe — APPROVED** [review]

File: Carbon.Permissions/Functions/Get-CPermission.ps1 (hunk 4 above)
Comment (splatteredbits) [reply Carbon.Permissions/Functions/Get-CPermission.ps1:70]: No. Cross PowerShell, but not cross-platform.

File: Carbon.Permissions/Functions/Test-CPermission.ps1 (hunk 5 above)
Comment (splatteredbits) [reply Carbon.Permissions/Functions/Test-CPermission.ps1]: I was debating whether or not to rename this. Debate closed. Renamed.

File: README.md (hunk 9 above)
Comment (splatteredbits) [reply README.md]: It doesn't work on PowerShell 6. Updated to be PowerShell 7+.

**Pull request merged**